src/service/navigation.js

```diff
@@ -738,6 +742,29 @@ export class Navigation {
       getMode().test && !this.isEmbed_ ? this.ampdoc.win.location.href : '';
     return this.parseUrl_(baseHref);
   }
+
+  /**
+   * Requests navigation through a Viewer to the given destination. If the viewer does
+   * not support this operation, does nothing.
+   * @param {string} url A URL.
+   * @param {string} requestedBy Informational string about the entity that
+   *     requested the navigation.
+   * @return {boolean} Returns true if navigation message was sent to viewer.
+   *     Otherwise, returns false.
+   */
+  navigateToUrlWithViewer(url, requestedBy) {
+    if (this.viewer_.hasCapability('interceptNavigation')) {
```

Comment:
Generally, more powerful features like this require the viewer to be "[trusted](https://github.com/ampproject/amphtml/blob/875ca95a4ef8071c0e8ae465803b179756770fc8/src/service/viewer-interface.js#L141-L146)", meaning we have some relationship with the viewer's origin and the origin has been [whitelisted](https://github.com/ampproject/amphtml/blob/875ca95a4ef8071c0e8ae465803b179756770fc8/src/config.js#L83-L95).

/cc @choumx

Reply:
+1. For XHR interception in email, we also require the document to "opt itself in" via an attribute on the `<html>` tag. This attribute is also invalid AMP (and thus injected by the viewer), which implies that the page must be on a special embedding surface. See `allow-xhr-interception` in the codebase.

Reply:
Interesting - do you want me to add a new attribute, say "allow-navigation-interception", and check for that here along with the 'trusted viewer' and capability check?

Reply:
Yes, please follow that convention.

Reply:
Ok I updated the checks to verify the viewer is either trusted or localhost, and also added a check for an "allow-navigation-interception" attribute on the <html> tag. Please take a look when you get a chance. :)